paasta_tools/kubernetes/remote_run.py

```diff
@@ -0,0 +1,373 @@
+    """
+    kube_client = KubeClient(config_file="/etc/kubernetes/admin.conf")
+
+    # Load the service deployment settings
+    deployment_config = agnostic_load_service_config(service, instance, cluster)
+    namespace = deployment_config.get_namespace()
+
+    # Rebuild the job metadata
+    formatted_job = deployment_config.format_kubernetes_job(job_label="remote-run")
+    job_name = _format_remote_run_job_name(formatted_job, user)
+
+    # Find pod and create exec token for it
+    pod = find_pod(kube_client, namespace, job_name)
+    if not pod:
+        raise RemoteRunError(f"Pod for {job_name} not found")
+    pod_name = pod.metadata.name
+    logger.info(f"Generating temporary service account token for {pod_name}")
+    service_account = create_remote_run_service_account(
+        kube_client, namespace, pod_name, user
+    )
+    role = create_pod_scoped_role(kube_client, namespace, pod_name, user)
+    bind_role_to_service_account(kube_client, namespace, service_account, role)
+    return create_temp_exec_token(kube_client, namespace, service_account)
+
+
+def find_pod(
+    kube_client: KubeClient, namespace: str, job_name: str, retries: int = 3
+) -> Optional[V1Pod]:
+    """Locate pod for remote-run job
+
+    :param KubeClient kube_client: Kubernetes client
+    :param str namespace: the pod namespace
+    :param str job_name: remote-run job name
+    :param int retries: maximum number of attemps
+    :return: pod object if found
+    """
+    for _ in range(retries):
+        pod_list = kube_client.core.list_namespaced_pod(namespace)
+        for pod in pod_list.items:
+            if pod.metadata.name.startswith(job_name):
+                return pod
+        sleep(0.5)
+    return None
+
+
+def create_temp_exec_token(
+    kube_client: KubeClient,
+    namespace: str,
+    service_account: str,
+) -> str:
+    """Create a short lived token for service account
+
+    :param KubeClient kube_client: Kubernetes client
+    :param str namespace: service account namespace
+    :param str service_account: service account name
+    :return: token value
+    """
+    token_spec = V1TokenRequestSpec(
+        expiration_seconds=600,  # minimum allowed by k8s
+        audiences=[],
+    )
+    request = AuthenticationV1TokenRequest(spec=token_spec)
+    response = kube_client.core.create_namespaced_service_account_token(
+        service_account, namespace, request
+    )
+    return response.status.token
+
+
+def create_remote_run_service_account(
+    kube_client: KubeClient,
+    namespace: str,
+    pod_name: str,
+    user: str,
+) -> str:
+    """Create service account to exec into remote-run pod
+
+    :param KubeClient kube_client: Kubernetes client
+    :param str namespace: pod namespace
+    :param str pod_name: pod name
+    :param str user: user requiring credentials
+    """
+    pod_name_hash = hashlib.sha1(pod_name.encode("utf-8")).hexdigest()[:12]
+    service_account_name = f"remote-run-{user}-{pod_name_hash}"
```

Comment:
i don't think we're really in danger of hitting the length limit for SA names, but might be good to future-proof this by wrapping thie with our truncation function just in case luispwithasuperduperreallylonghowlongisthisihopeitsover250somethingcharacters is ever hired :p

Reply:
I love DNS